### PR TITLE
refactor material banner and fix focus issue

### PR DESCRIPTION
**Pull request by DBowen33**
- Refactored material banner setup to make it more simple and stateless
- Added focusNode to where the dismiss button focuses automatically when banner is shown

Before: https://screencast.googleplex.com/cast/NTI4OTMxNjU2MDQwNDQ4MHw2ODdiMmYwOC1hMg
After: https://screencast.googleplex.com/cast/NjAzNTY4NDA2NTI4MDAwMHwyNTRjMWUxMi0zMA

fixes b/346646604

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Review by hannah-hyj — APPROVED** [review]
LGTM

File: dev/a11y_assessments/test/material_banner_test.dart [hunk 1]
```diff
@@ -20,4 +21,13 @@ void main() {
     await tester.pumpAndSettle();
     expect(find.text('Hello, I am a Material Banner'), findsNothing);
   });
+
+  testWidgets('dismiss button focused on banner open', (WidgetTester tester) async {
+    await pumpsUseCase(tester, MaterialBannerUseCase());
+    await tester.tap(find.text('Show a MaterialBanner'));
+    await tester.pumpAndSettle();
+
+    final TextButton dismissButtonFinder = tester.widget(find.byType(TextButton)) as TextButton;
```
Comment (navaronbracke) [inline dev/a11y_assessments/test/material_banner_test.dart:30]:
```suggestion
    final TextButton dismissButtonFinder = tester.widget<TextButton>(find.byType(TextButton));
```

File: dev/a11y_assessments/lib/use_cases/material_banner.dart [hunk 2]
```diff
@@ -15,57 +15,44 @@ class MaterialBannerUseCase extends UseCase {
...
+  MainWidget({super.key});
 
-  @override
-  State<MainWidget> createState() => MainWidgetState();
-}
-
-class MainWidgetState extends State<MainWidget> {
-  double currentSliderValue = 20;
-  ScaffoldFeatureController<MaterialBanner, MaterialBannerClosedReason>? controller;
+  final FocusNode dismissButtonFocusNode = FocusNode();
```
Comment (navaronbracke) [inline dev/a11y_assessments/lib/use_cases/material_banner.dart:24]: This focus node is never disposed. You'll have to keep the `MainWidget` stateful so you can dispose the focus node in the dispose method of that widget's State class.

You can also keep the `MainWidget` constructor const.

Comment (DBowen33) [reply dev/a11y_assessments/lib/use_cases/material_banner.dart]: Thanks for the review! Added the state back to the widget along adding another focus node to put focus back on the show button whenever dismiss is pressed.

However, I'm running into a weird issue that I can't figure out:
[Screen recording 2024-08-02 1.52.03 PM.webm](https://github.com/user-attachments/assets/0125b4d3-cb19-408d-985e-af665e2aea05)

When clicking the show button and dismiss button, the behavior works fine, but whenever I press Enter on the dismiss button, I keep getting an assertion error with InkWell saying "InkResponse must be done with layout before starting a splash.". This only happens when pressing Enter on the dismiss button, and I can't quite figure out what the issue is. Any suggestions would be greatly appreciated!

File: dev/a11y_assessments/lib/use_cases/material_banner.dart [hunk 3]
```diff
@@ -24,48 +24,62 @@ class MainWidget extends StatefulWidget {
   @override
   State<MainWidget> createState() => MainWidgetState();
 }
-
 class MainWidgetState extends State<MainWidget> {
-  double currentSliderValue = 20;
-  ScaffoldFeatureController<MaterialBanner, MaterialBannerClosedReason>? controller;
+
+  FocusNode dismissButtonFocusNode = FocusNode();
```
Comment (navaronbracke) [inline dev/a11y_assessments/lib/use_cases/material_banner.dart:29]:
```suggestion
  final FocusNode dismissButtonFocusNode = FocusNode();
  final FocusNode showButtonFocusNode = FocusNode();
```

File: dev/a11y_assessments/lib/use_cases/material_banner.dart [hunk 4]
```diff
@@ -24,48 +24,62 @@ class MainWidget extends StatefulWidget {
...
+      MaterialBanner(
+        padding: const EdgeInsets.all(20),
+        content: const Text('Hello, I am a Material Banner'),
+        leading: const Icon(Icons.agriculture_outlined),
+        backgroundColor: Colors.blue,
+        actions: <Widget>[
+          TextButton(
+            focusNode: dismissButtonFocusNode,
+            onPressed: () {
+              setState(() {
```
Comment (navaronbracke) [inline dev/a11y_assessments/lib/use_cases/material_banner.dart:55]: This call to `setState` is not needed, as no state is updated. You could also use a function tear-off here for onPressed.

```suggestion
TextButton(
  focusNode: dismissButtonFocusNode,
  onPressed: hideBanner,
  child: const Text('DISMISS'),
)
```

File: dev/a11y_assessments/lib/use_cases/material_banner.dart [hunk 5]
```diff
@@ -24,48 +24,62 @@ class MainWidget extends StatefulWidget {
...
-    }
     return Scaffold(
       appBar: AppBar(
         backgroundColor: Theme.of(context).colorScheme.inversePrimary,
         title: const Text('MaterialBanner'),
       ),
       body: Center(
         child: ElevatedButton(
-          onPressed: onPress,
+          focusNode: showButtonFocusNode,
```
Comment (navaronbracke) [inline dev/a11y_assessments/lib/use_cases/material_banner.dart:76]: Same here, the setState is not needed.
Also, the `child` should go last per https://dart.dev/tools/linter-rules/sort_child_properties_last

```suggestion
ElevatedButton(
  focusNode: showButtonFocusNode,
  onPressed: showBanner,
  child: const Text('Show a MaterialBanner'),
)
```

**Review by navaronbracke — COMMENTED** [review]
Some minor remarks still, but I'm not sure why that assertion is being thrown, though.

**Comment by DBowen33** [comment]
> Some minor remarks still, but I'm not sure why that assertion is being thrown, though.

Looks like the assertion error had to do with setting the state, now that I've taken the set state out the assertion is not being thrown anymore. Thank you!

File: dev/a11y_assessments/lib/use_cases/material_banner.dart [hunk 6]
```diff
@@ -24,50 +24,56 @@ class MainWidget extends StatefulWidget {
...
         title: const Text('MaterialBanner'),
       ),
       body: Center(
         child: ElevatedButton(
-          onPressed: onPress,
-          child: const Text('Show a MaterialBanner'),
+          focusNode: showButtonFocusNode,
+          onPressed: showBanner,
+          child: const Text('Show a MaterialBanner')
+          )
```
Comment (navaronbracke) [inline dev/a11y_assessments/lib/use_cases/material_banner.dart:75]: The formatting here isn't entirely correct, due to missing commas. (after the `child` and after the closing parens of the `ElevatedButton`)
Be aware, though that flutter/flutter does not use `dart format` (yet) for formatting in the repo, per https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#formatting

So you might need to do some hand formatting here.

```suggestion
ElevatedButton(
  focusNode: showButtonFocusNode,
  onPressed: showBanner,
  child: const Text('Show a MaterialBanner'),
),
```

**Review by navaronbracke — COMMENTED** [review]
Just some missing commas and formatting for the MaterialBanner use case file, but otherwise this looks good to me.

Unfortunately I'm unable to give you an approval myself, but I'm going to reach out to get an additional reviewer.

File: dev/a11y_assessments/lib/use_cases/material_banner.dart (hunk 6 above)
Comment (DBowen33) [reply dev/a11y_assessments/lib/use_cases/material_banner.dart]: Done! Thank you for the feedback!

**Review by chunhtai — COMMENTED** [review]
Sorry I realized I have not send the review comment

File: dev/a11y_assessments/lib/use_cases/material_banner.dart [hunk 7]
```diff
@@ -24,47 +24,53 @@ class MainWidget extends StatefulWidget {
   @override
   State<MainWidget> createState() => MainWidgetState();
 }
-
```
[the remark this answers is not in the document]
Comment (DBowen33) [reply dev/a11y_assessments/lib/use_cases/material_banner.dart:27]: Done!

**Review by chunhtai — APPROVED** [review]
LGTM, looks like you need to rebase to fix the conflict and ci

**Pull request merged**